modules/openxr/extensions/openxr_fb_scene_capture_extension_wrapper.cpp

```diff
@@ -0,0 +1,112 @@
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "openxr_fb_scene_capture_extension_wrapper.h"
+#include "modules/openxr/openxr_interface.h"
+
+using namespace godot;
+
+OpenXRFbSceneCaptureExtensionWrapper *OpenXRFbSceneCaptureExtensionWrapper::singleton = nullptr;
+
+OpenXRFbSceneCaptureExtensionWrapper *OpenXRFbSceneCaptureExtensionWrapper::get_singleton() {
+	return singleton;
+}
+
+OpenXRFbSceneCaptureExtensionWrapper::OpenXRFbSceneCaptureExtensionWrapper(OpenXRAPI *p_openxr_api) :
+		OpenXRExtensionWrapper(p_openxr_api) {
+	request_extensions[XR_FB_SCENE_CAPTURE_EXTENSION_NAME] = &fb_scene_capture_ext;
+	singleton = this;
+}
+
+OpenXRFbSceneCaptureExtensionWrapper::~OpenXRFbSceneCaptureExtensionWrapper() {
+	cleanup();
+}
+
+void OpenXRFbSceneCaptureExtensionWrapper::cleanup() {
+	fb_scene_capture_ext = false;
+}
+
+void OpenXRFbSceneCaptureExtensionWrapper::on_instance_created(const XrInstance instance) {
+	if (fb_scene_capture_ext) {
+		bool result = initialize_fb_scene_capture_extension(instance);
+		if (!result) {
+			print_error("Failed to initialize fb_scene_capture extension");
+			fb_scene_capture_ext = false;
+		}
+	}
+}
+
+void OpenXRFbSceneCaptureExtensionWrapper::on_instance_destroyed() {
+	cleanup();
+}
+
+bool OpenXRFbSceneCaptureExtensionWrapper::request_scene_capture() {
+	XrAsyncRequestIdFB requestId;
+	XrSceneCaptureRequestInfoFB request;
+	request.type = XR_TYPE_SCENE_CAPTURE_REQUEST_INFO_FB;
+	request.next = nullptr;
+	request.requestByteCount = 0;
+	request.request = nullptr;
+	XrResult result = xrRequestSceneCaptureFB(openxr_api->get_session(), &request, &requestId);
+	scene_capture_enabled = (result == XR_SUCCESS);
+	return scene_capture_enabled;
+}
+
+bool OpenXRFbSceneCaptureExtensionWrapper::is_scene_capture_enabled() {
+	return scene_capture_enabled;
+}
+
+bool OpenXRFbSceneCaptureExtensionWrapper::initialize_fb_scene_capture_extension(const XrInstance p_instance) {
+	ERR_FAIL_NULL_V(openxr_api, false);
+
+	EXT_INIT_XR_FUNC_V(xrRequestSceneCaptureFB);
+
+	return true;
+}
+
+VSet<XrStructureType> OpenXRFbSceneCaptureExtensionWrapper::get_handled_events() {
+	VSet<XrStructureType> events;
+	events.insert(XR_TYPE_EVENT_DATA_SCENE_CAPTURE_COMPLETE_FB);
+	return events;
+}
+
+void OpenXRFbSceneCaptureExtensionWrapper::handle_event(XrEventDataBuffer event) {
+	ERR_FAIL_NULL(openxr_api);
+
+	if (event.type == XR_TYPE_EVENT_DATA_SCENE_CAPTURE_COMPLETE_FB) {
+		scene_capture_enabled = false;
+		OpenXRInterface *xr_interface = openxr_api->get_xr_interface();
+		if (xr_interface) {
+			xr_interface->on_scene_capture_completed();
```

Comment:
I don't think I like this calling back into the XR interface, also what happens on `on_scene_capture_completed`? If I understand correctly it sends a signal back into Godot but what is the user then supposed to do?

In ARKit as we get information of our surroundings we started to register new XRPositionalTracker objects to the XRServer which could be picked up by XRAnchor3D objects. There are holes here because ARKit at the time didn't go very far with scene understanding yet.

I'll have to look a little more into the API FB created here to wrap my head around what the best way forward is.

Reply:
This API essentially requests the Quest to take over and opens a new app to allow the user the mark the different objects in its surroundings. There is no automated recognition of anything. When the screen capture complete signal is emitted, then we can query the spatial anchors using the Spatial Anchor API (@konczg is working on that now) that were created during the screen capture process.